ludwig/models/llm.py

```diff
@@ -33,16 +34,40 @@ def __init__(
         self,
         config_obj: LLMModelConfig,
         random_seed=None,
+        device=None,
         **_kwargs,
     ):
         super().__init__(random_seed=random_seed)
 
         self.config_obj = config_obj
         self._random_seed = random_seed
 
+        self.model_name = self.config_obj.model_name
+
+        logger.info("Loading large language model...")
         self.model = AutoModelForCausalLM.from_pretrained(self.config_obj.model_name)
-        self.generation_config = GenerationConfig(**self.config_obj.generation_config.to_dict())
-        self.max_new_tokens = self.config_obj.generation_config.max_new_tokens
+        self.curr_device = torch.device("cpu")  # model initially loaded onto cpu
+        logger.info("Done.")
+
+        if device is None:
+            device = get_torch_device()
+        self.to_device(device)
+
+        self.max_new_tokens = self.config_obj.generation.max_new_tokens
+
+        # Determines the maximum length of the context (input + output tokens)
+        if hasattr(self.model.config, "max_sequence_length"):
+            self.context_len = self.model.config.max_sequence_length
+        elif hasattr(self.model.config, "max_position_embeddings"):
+            self.context_len = self.model.config.max_position_embeddings
+        else:
+            self.context_len = 2048
+        self.max_input_length = self.context_len - self.max_new_tokens - 8
```

Comment:
Worth adding a comment about where this logic came from (FastChat), or explain why we subtract 8.

Reply:
Done